src/java/org/apache/cassandra/io/sstable/SSTableWatcher.java

```diff
@@ -53,4 +65,43 @@ default Set<Component> discoverComponents(Descriptor descriptor, Set<Component>
     {
         return existing;
     }
+
+    /**
+     * Appends new component names to the TOC component.
+     */
+    default void appendTOC(Descriptor descriptor, Collection<Component> components)
+    {
+        File tocFile = descriptor.fileFor(Component.TOC);
+        writeTOC(tocFile, components, APPEND);
+    }
+
+    /**
+     * Write TOC file with given components and write mode
+     */
+    default void writeTOC(File tocFile, Collection<Component> components, File.WriteMode writeMode)
+    {
+        FileOutputStreamPlus fos = null;
+        try (PrintWriter w = new PrintWriter((fos = tocFile.newOutputStream(writeMode))))
+        {
+            for (Component component : components)
+                w.println(component.name);
+            w.flush();
+            fos.sync();
+        }
+        catch (IOException e)
+        {
+            throw new FSWriteError(e, tocFile);
+        }
+    }
+
+    /**
+     * Rewrite TOC components with given components. The default implementation deletes the existing file and append new components
+     */
+    default void rewriteTOC(Descriptor descriptor, Collection<Component> components)
+    {
+        File tocFile = descriptor.fileFor(Component.TOC);
```

Comment:
I know this is existing code, but it's weird that if we fail to delete the toc, we log it but otherwise "APPEND" to the TOC, which is the wrong thing to do and will create a broken TOC that will probably generate other future errors.

Since you helpfully created a `writeTOC(..., WriteMode)`, why not just call `writeTOC(tocFile, components, OVERWRITE)` and let any error be thrown as `FSWriteError`, which seems overall the "more correct" behaviour?